src/librustc_incremental/persist/save.rs

```diff
@@ -98,38 +101,37 @@ fn save_in<F>(sess: &Session, opt_path_buf: Option<PathBuf>, encode: F)
     }
 }
 
-pub fn encode_dep_graph<'a, 'tcx>(hcx: &mut HashContext<'a, 'tcx>,
-                                  builder: &mut DefIdDirectoryBuilder,
-                                  query: &DepGraphQuery<DefId>,
-                                  encoder: &mut Encoder)
-                                  -> io::Result<()> {
-    let (nodes, edges) = (query.nodes(), query.edges());
-
-    // Create hashes for inputs.
-    let hashes = nodes.iter()
-        .filter_map(|dep_node| {
-            hcx.hash(dep_node)
-                .map(|(_, hash)| {
-                    let node = builder.map(dep_node);
-                    SerializedHash {
-                        node: node,
-                        hash: hash,
-                    }
-                })
-        })
-        .collect();
+pub fn encode_dep_graph(preds: &Predecessors,
+                        builder: &mut DefIdDirectoryBuilder,
+                        encoder: &mut Encoder)
+                        -> io::Result<()> {
+    // Create a flat list of (Input, WorkProduct) edges for
+    // serialization.
+    let mut edges = vec![];
+    for (&target, sources) in &preds.inputs {
+        match *target {
+            DepNode::MetaData(_) => continue, // see encode_metadata_hashes instead
```

Comment:
Oh right, it's just matching on the target. Nevermind then.
Regarding a separate `DepNode` variant: That would make sense. Also, with some additional machinery, similar to what we do for object files, we could integrate the metadata-object into the dependency graph too and avoid rewriting it if nothing has changed (which it shouldn't need to if there are only changes in unexported functions).